### PR TITLE
Remove traversal strategy

### PR DESCRIPTION
**Pull request by bakura10**
ping @danizord 

The user can still customize the strategy if he wants by providing a subclass for Rbac and overrinding getRolesIterator. But for most people, it simplifies the code.

(tests not updated yet)

### CODE REVIEW

**Comment by danizord** [comment]
Yeah, now that we are 5.5+ we can always use generators to handle the traversal. But I still think that we could keep the traversal strategy concept and provide at least 2 built-in strategies:

* `FlatTraversalStrategy`
* `HierarchicalTraversalStrategy`

Mainly because I have some projects that does not use hierarchical roles. BTW I always prefer having users with multiple roles than users with hierarchical roles.

What do you think?

**Comment by bakura10** [comment]
I definitely prefer having multiple roles too and I've never used the hierarchical. I'm just asking if we have any interest in having a separate concept. Not only it forces us to define a factory to create the Rbac object but I'm not sure it is really used. And people can still extend the rbac if they want to change the logic traversal. No?

**Comment by danizord** [comment]
The point is that the current traversal algorithm brings a bit of overhead for people who does not use hierarchical roles.

>Not only it forces us to define a factory to create the Rbac object but I'm not sure it is really used

We could provide static factory methods to create Rbac instances with built-in traversal strategies:

```php
$rbac = Rbac::fromFlatTraversalStrategy();
$rbac = Rbac::fromHierarchicalTraversalStrategy();
```

But you're right. It's not worth introducing that complexity just to avoid a very small overhead. So, I'm ok to merge it ;-)

File: src/Rbac.php [hunk 1]
```diff
@@ -57,12 +43,23 @@ public function isGranted($roles, $permission)
...
 
     /**
-     * Get the strategy.
-     *
-     * @return TraversalStrategyInterface
+     * @param  RoleInterface|RoleInterface[]|Traversable $roles
+     * @return Generator
      */
-    public function getTraversalStrategy()
+    protected function getRolesIterator($roles)
```
Comment (danizord) [inline src/Rbac.php]: Maybe `flattenRoles()` sounds a more descriptive name?

File: src/Rbac.php [hunk 2]
```diff
@@ -57,12 +43,23 @@ public function isGranted($roles, $permission)
     }
 
     /**
-     * Get the strategy.
-     *
-     * @return TraversalStrategyInterface
+     * @param  RoleInterface|RoleInterface[]|Traversable $roles
```
Comment (danizord) [inline src/Rbac.php]: Only traversables are accepted here:  `@param  RoleInterface[]|Traversable $roles`

Comment (bakura10) [inline src/Rbac.php]: no it also accepts the RoleInterface (and converts it to array).

Comment (danizord) [inline src/Rbac.php]: rsz_59792059

Comment (bakura10) [inline src/Rbac.php]: Sure: https://github.com/zf-fr/rbac/blob/simplfy/src/Rbac.php#L32 :D

Comment (danizord) [inline src/Rbac.php]: But this docblock is for `getRolesIterator()` method.

Comment (bakura10) [inline src/Rbac.php]: Oh :D... I'm updating the code right now.

**Comment by bakura10** [comment]
Done @danizord , tests are now passing. I'd therefore like to take 2.0 so this can be the base of the ZfcRbac. Is there anything else you'd like to change/see in this library?

**Comment by danizord** [comment]
@bakura10 Can we force `$permission` type to be string everywhere? I'd like to avoid `mixed` but that breaks BC

**Comment by bakura10** [comment]
I don't really care about BC here. However I think it would limit things here. Maybe your permission could be an object that would contain additional info (like a context...) ?

**Comment by danizord** [comment]
@bakura10 yeah that's the problem. Previous version of ZfcRbac people was trying to overuse RBAC for complex domain logic. I'd prefer to be specific here - let ZfcRbac work with objects and typecast to string before passing to RBAC component

**Comment by bakura10** [comment]
I mean, the fact that Rbac just pass the permission to the role makes me think that Rbac component actually does not bother. What kind of advantage do we have of enforcing a string as permission, while we can give the flexibility to do the logic there ? In some complex situation, even with permission, you may be interested in wrapping your permission into an object that would contain the context or any useful information that may be of interest. No?

**Comment by bakura10** [comment]
I've added code to enforce string for permission. Actually you're right, because the context is given as a separate argument in ZfcRbac.

**Comment by Ocramius** [comment]
Just barely following the discussion, but remember that there is also `yield from` ;-)

**Comment by bakura10** [comment]
Isn't that PHP7?

**Comment by danizord** [comment]
@bakura10 considering "context" stuff is not actually "role based access control". IMHO stuff that is not checking if a role has a given permission should happen either before or after calling RBAC.

@Ocramius PHP 7 only =\

**Comment by bakura10** [comment]
Okay... so... ready for merging that and tagging v2 ? :D

**Comment by danizord** [comment]
Yup

**Comment by Ocramius** [comment]
> Isn't that PHP7?

Yup, which is the current stable release as of this coming week ;-)

**Comment by bakura10** [comment]
Library has been tagged to v2. Thanks everyone, let's move on to ZfcRbac now :).

**Comment by danizord** [comment]
> Yup, which is the current stable release as of this coming week ;-)

But we aim ZF3 here (5.5+). BTW I heard somewhere that ZF3 would have a PHP7 branch with scalar/return typehints and such, do you have some info about that?

**Comment by bakura10** [comment]
ZF3 has not really much meaning to me anymore...

**Comment by Ocramius** [comment]
> Library has been tagged to v2. Thanks everyone, let's move on to ZfcRbac now :)

Awesome :-)

> do you have some info about that?

Nope 

> ZF3 has not really much meaning to me anymore...

Pretty much agree - make small libs, keep iterating on those (rather than on a massive aggregate lib): we failed enough times at that :-P

Kudos for the release =)

**Comment by bakura10** [comment]
That was the super easy part. ZfcRbac is now a much bigger task... =)

**Pull request merged**